cloud_functions/create_spritesheet.py

```diff
@@ -0,0 +1,259 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Used to create a spritesheet from the given thumbnails.
+
+This Cloud Function runs to create a spritesheet of all the thumbnails
+that have been uploaded to the bucket. The function checks to make sure
+that the file isn't a embeddings or the spritesheet itself, and concatanates
+all thumbnails into a numpy array
+"""
+
+import os
+import tempfile
+import base64
+from google.cloud import storage, vision, datastore
+import json
+import re
+import sys
+import numpy as np
+import cv2
+import json
+
+# Initiating the storage & datastore clients.
+storage_client = storage.Client()
+datastore_client = datastore.Client('step-2020-johndallard')
+vision_client = vision.ImageAnnotatorClient()
+bucket_name = 'embeddings_visualizer_output_bucket'
+spritesheet_folder_name = 'spritesheets'
+SPRITESHEET = 'spritesheet'
+
+
+def create_spritesheet(file_data, context):
+    """Creates a spritesheet.
+
+    Retrieves thumbnails and outputs them to BUCKET_NAME in a spritesheet
+    using a numpy array. Images will be read using cv2
+
+    Args:
+        file_data:
+            Dictionary that contains data specific to the event.
+        context:
+            Cloud functions eventmetadata.
+
+    Returns:
+        A dict mapping keys to the corresponding table row data
+        fetched. Each row is represented as a tuple of strings. For
+        example:
+
+        {b'Serak': ('Rigel VII', 'Preparer'),
+         b'Zim': ('Irk', 'Invader'),
+         b'Lrrr': ('Omicron Persei 8', 'Emperor')}
+
+        Returned keys are always bytes.  If a key from the keys argument is
+        missing from the dictionary, then that row was not found in the
+        table (and require_all_keys must have been False).
+    """
+
+    file_name = file_data['name']
+    user_name, dataset_name = get_user_and_dataset_name(file_name)
+
+    spritesheet_file_name = (os.path.join(user_name, dataset_name,
+                                          spritesheet_folder_name, SPRITESHEET))
+
+    # We check to make sure the image is does not have the same name as the
+    # spritesheet, so that when we upload the spritesheet the function
+    # doesn't run again
+    if SPRITESHEET in file_name or '.npy' in file_name:
+        return
+    spritesheet_file_name = spritesheet_file_name + '.png'
+
+    # Check to make sure there is metadata, if none return no data and error 
+    # message
+    metadata = get_metadata(user_name, dataset_name)
+    if metadata is None:
+        return
+
+    num_expected_thumbnails = get_expected_thumbnail_number(metadata)
+
+    # Check the amount of images already in the bucket from the dataset
+    folder = os.path.join('/', file_name.split('/')[:-1])
```

Comment:
?? I don't understand what you did. Can you please re-do it? Please read all the APIs in os.path and find what you can use. Please do that for all the file name handlings everywhere.
The original code (your invented method to get the directory name) was smart, but you have to try to find and reuse existing libraries whenever possible. Why? They are better tested for edge cases. For example, some older file systems use '\' instead of '/' for a separator. Will your code work in that case?